cmd/launcher/doctor.go

```diff
@@ -0,0 +1,625 @@
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/kit/logutil"
+	"github.com/kolide/kit/version"
+	"github.com/kolide/launcher/pkg/agent/flags"
+	"github.com/kolide/launcher/pkg/agent/knapsack"
+	"github.com/kolide/launcher/pkg/agent/types"
+	"github.com/kolide/launcher/pkg/autoupdate"
+	"github.com/kolide/launcher/pkg/autoupdate/tuf"
+	"github.com/kolide/launcher/pkg/launcher"
+	"github.com/kolide/launcher/pkg/log/checkpoint"
+	"github.com/peterbourgon/ff/v3"
+	"github.com/shirou/gopsutil/v3/process"
+
+	"github.com/fatih/color"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	// Command line colors
+	cyanText   = color.New(color.FgCyan)
+	headerText = color.New(color.FgHiWhite).Add(color.Bold)
+	yellowText = color.New(color.FgHiYellow)
+
+	// Println functions for checkup details
+	whiteText = color.New(color.FgWhite)
+	greenText = color.New(color.FgGreen).PrintlnFunc()
+	redText   = color.New(color.FgRed).Add(color.Bold).PrintlnFunc()
+
+	// Indented output for checkup results
+	info = func(a ...interface{}) {
+		whiteText.Println(fmt.Sprintf("    %s", a...))
+	}
+	warn = func(a ...interface{}) {
+		yellowText.Println(fmt.Sprintf("    %s", a...))
+	}
+	fail = func(a ...interface{}) {
+		whiteText.Println(fmt.Sprintf("❌  %s", a...))
+	}
+	pass = func(a ...interface{}) {
+		whiteText.Println(fmt.Sprintf("✅  %s", a...))
+	}
+
+	configFile string
+	binDir     string
+)
+
+// checkup encapsulates a launcher health checkup
+type checkup struct {
+	name  string
+	check func() (string, error)
```

Comment:
This signature feels very simple, maybe a good enough start. But I think there are a couple things I don't know how to do. (but maybe they don't matter)

First, I don't think we can differentiate between an unexpected failure in code, vs code works but a broken condition. Like, if we failed to open a boltdb connection that feels different than if some piece of data isn't what we need. Both are the same failure to the user, but one feels like a protocol error, and the other feels like the correct reporting of data.

Second, I'm not sure how to represent warnings here. 

If I combine those, I wonder about `func() (info string, status enum, error)`

Reply:
But that suggestion makes it hard to differentiate the output for flare vs doctor.

Reply:
Oh, I see now. It's in the check's run call. hmmmmm